### DISCUSSION TITLE
Support the new "function calling" in chat completions

### DISCUSSION THREAD

**Discussion by syntax42** [betalgo/openai#291] [Ideas]
Just launched by OpenAI. The "function calling" method would be very usefull to get back structured data. Read more here: https://platform.openai.com/docs/guides/gpt/function-calling

**Comment by oferavnery** [comment]
+1. I wonder how streaming is going to work with the new function role and messages.

**Reply by syntax42** [reply]
I suspect "not well" or maybe not relevant as you would only want to take action, when the whole response is received.

**Reply by oferavnery** [reply]
Yep - the thing is that you can't know for sure if the response will include message content or a function call, so there should be a way to figure out it out pretty quickly, so if no function call is required you can still stream the response message.

**Reply by syntax42** [reply]
I think you will be guarantied a return in the form of json, or?

**Reply by oferavnery** [reply]
Not if the model thinks that no function call is needed and it returns a standard response. You can enforce calling a specific function but it kind of defeats the purpose of automatic tool selection.

**Comment by syntax42** [comment]
To clearify. I do not think it would be useful to make the c# SDK directly call functions. As far as i can see, that is primarily for execution in javascript. But maybe the "guarantied" structured response would be easier to parse into a c# object. Maybe as an utility function?

**Reply by oferavnery** [reply]
It would be (and still is ) useful for us - our entire OpenAI API processing is done server side with C# and streamed to Angular UI with SignalR.

**Comment by syntax42** [comment]
True

tor. 15. jun. 2023 09.03 skrev oferavnery ***@***.***>:
